lib/internal/worker/locks.js

```diff
@@ -0,0 +1,526 @@
+/* global SharedArrayBuffer */
+'use strict';
+const {
+  broadcastChannel,
+  DOMException
+} = internalBinding('messaging');
+const {
+  threadId,
+  isMainThread,
+} = require('internal/worker');
+const {
+  receiveMessageOnPort
+} = require('internal/worker/io');
+const {
+  sticky_message_symbol,
+  onclose_message_symbol,
+} = internalBinding('symbols');
+const assert = require('internal/assert');
+const {
+  SafeMap,
+  SymbolToStringTag,
+  Symbol,
+  BigUint64Array,
+  Int32Array,
+  Promise,
+  ObjectCreate,
+  PromiseResolve,
+  ObjectDefineProperties,
+  TypeError,
+} = primordials;
+
+const kName = Symbol('kName');
+const kMode = Symbol('kMode');
+
+// Class that contains all the state around active and pending locks.
+class LockInternals {
+  constructor() {
+    // We identify each thread by the pid + thread id combination.
+    this.clientId = `node-${process.pid}-${threadId}`;
+
+    // These maps store the resolve/reject functions for a) when the lock
+    // request is granted and b) when the final result of the operation is
+    // known.
+    this.availableMap = new SafeMap();
+    this.resolveMap = new SafeMap();
+    this.queueProcessingScheduled = false;
+
+    // Create an internal BroadcastChannel instance for communication with
+    // other threads.
+    this.channel = broadcastChannel('nodejs-internal:locking');
+    this.channel.on('message', (msg) => this.onmessage(msg));
+    // Post the message that will be sent when this thread closes down
+    // (regardless of how exactly that happens).
+    this.channel.postMessage({
+      op: 'terminated',
+      clientId: this.clientId,
+      threadId: threadId,
+      [onclose_message_symbol]: true
+    });
+    // By default, do not keep the event loop running just for locks.
+    this.channel.unref();
+
+    // Look for existing messages on the channel. There should be some, unless
+    // we are the first thread to set this up.
+    this.readSynchronousMessages();
+    if (!this.state) {
+      // This assertion should be valid because the first Worker instantiation
+      // on the main thread makes sure that locks will be initialized.
+      assert(isMainThread);
+      const sab = new SharedArrayBuffer(16);
+      this.state = {
```

Comment:
This is _kind of_ confusing but I can't think of a nicer way to maintain this state.